### PR TITLE
Add support for `tls` protocol

### PR DESCRIPTION
**Pull request by whereistejas**
Hi @mxinden,
Creating this PR to close the issue #47.
This PR refers the Go code from [PR 153](https://github.com/multiformats/go-multiaddr/pull/153).

i have a couple of questions regarding the tests, it seems like we are missing a few tests:
1. [`construct_success`](https://github.com/multiformats/go-multiaddr/blob/eef92462702954520ff59a5bab357babd1a55ce5/multiaddr_test.go#L170-L173)
2. [`TestTextMarshaler`](https://github.com/multiformats/go-multiaddr/blob/eef92462702954520ff59a5bab357babd1a55ce5/multiaddr_test.go#L654-L668)

### CODE REVIEW

File: src/protocol.rs [hunk 1]
```diff
@@ -14,6 +14,8 @@ use std::{
 use unsigned_varint::{encode, decode};
 use crate::onion_addr::Onion3Addr;
 
+// All these values are obtained by converting hexadecimal protocol codes to u32.
+// The protocol codes are present in multiformats/multicodec repository.
```
Comment (mxinden) [inline src/protocol.rs:18]: :+1: for documenting this.

I am not exactly sure what the source of truth is, though I think we should link to multiformats/multiaddr instead.

```suggestion
// Protocols as well as their corresponding codes are defined in
// https://github.com/multiformats/multiaddr/blob/master/protocols.csv .
```

**Review by mxinden — COMMENTED** [review]
Thanks @whereistejas for the help!

Could you extend the `<Protocol as Arbitrary>` implementation with the new `Protocol` `Tls`?

https://github.com/multiformats/rust-multiaddr/blob/95d5e6415f7149e25be90ca3c23be71f33edd115/tests/lib.rs#L86-L89

> construct_success

Feel free to add additional test cases.

> TestTextMarshaler

Would this not be covered by `to_from_str_identityt`?

**Comment by whereistejas** [comment]
Hi @mxinden , I have added the extra tests and added the `Tls` proto, too. Please, let me know if any other changes are pending/required. :)

**Review by mxinden — COMMENTED** [review]
As far as I can tell you would need to adjust the `Arbitrary` implementation of `Proto` to potentially generate the value `25`. The modulo operation is a bit error prone, though I don't know of a better solution.

```diff
-        match u8::arbitrary(g) % 25  { // TODO: Add Protocol::Quic
+        match u8::arbitrary(g) % 26  { // TODO: Add Protocol::Quic
```

**Comment by whereistejas** [comment]
@mxinden I have made the change you suggested.

**Review by mxinden — APPROVED** [review]
Wow, thank you for the quick turn-around time @whereistejas!

Would you mind including a changelog entry? Other than that, this looks good to me. Thanks.

```diff
diff --git a/CHANGELOG.md b/CHANGELOG.md
index 65e438e..42a6fb7 100644
--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,9 @@
+# 0.14.0 [unreleased]
+
+- Add support for `tls` protocol. See [PR 48].
+
+[PR 48]: https://github.com/multiformats/rust-multiaddr/pull/48
+
 # 0.13.0 [2021-07-08]
 
 - Update to multihash v0.14.0 (see [PR 44]).
diff --git a/Cargo.toml b/Cargo.toml
index 33cce4c..8162f62 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -7,7 +7,7 @@ keywords = ["multiaddr", "ipfs"]
 license = "MIT"
 name = "multiaddr"
 readme = "README.md"
-version = "0.13.0"
+version = "0.14.0"
```

**Comment by whereistejas** [comment]
Hi @mxinden, done!

**Pull request merged**